src/containers/Application/ApplicationPage.tsx

```diff
@@ -1,34 +1,98 @@
 
 interface ApplicationProps {
   structure: FullStructure
   responses?: ResponsesByCode
 }
 
+const getFirstInvalidPage = ( fullStructure: FullStructure ):  SectionAndPage | null => {
+  // TODO implement, should rely on .progress
+  // return { sectionCode: 'S1', pageName: 'Page 2' }
+  return null
+}
+
+interface RevalidationState {
+  methodToCallOnRevalidation: MethodToCallOnRevalidation | null
+  shouldProcessValidation: boolean
+  lastRevalidationRequest: number
+}
+
 const ApplicationPage: React.FC<ApplicationProps> = ({ structure }) => {
-  const [strictSectionPage, setStrictSectionPage] = useState({ section: null, page: null })
-  const { error, isLoading, fullStructure, responsesByCode } = useGetFullApplicationStructure({
-    structure,
-  })
   const {
     userState: { currentUser },
   } = useUserState()
   const { push, query } = useRouter()
 
+  const {
+    state: { isLastElementUpdateProcessed, elementUpdatedTimestamp },
+  } = useFormElementUpdateTracker()
+
+  const [strictSectionPage, setStrictSectionPage] = useState<SectionAndPage | null>(null)
+  const [revalidationState, setRevalidationState] = useState<RevalidationState>({
+    methodToCallOnRevalidation: null,
+    shouldProcessValidation: false,
+    lastRevalidationRequest: Date.now(),
+  })
+
+  const shouldRevalidate = isLastElementUpdateProcessed && revalidationState.shouldProcessValidation
+  const revalidateAfterTimestamp = shouldRevalidate ? elementUpdatedTimestamp : 0
+
+  const { error, fullStructure } = useGetFullApplicationStructure({
+    structure,
+    shouldRevalidate,
+    revalidateAfterTimestamp,
+  })
+
   const currentSection = query.sectionCode
   const currentPage = `Page ${query.page}`
 
+  /* Method to pass to progress bar, next button and submit button  to cause revalidation before aciton can be proceeded
+     Should always be called on submit, but only be called on next or progress bar navigation when isLinear */
+  // TODO may rename if we want to display loading modal ?
+  const requestRevalidation = (methodToCall: MethodToCallOnRevalidation) => {
+    setRevalidationState({
+      methodToCallOnRevalidation: methodToCall,
+      shouldProcessValidation: true,
+      lastRevalidationRequest: Date.now(),
+    })
+    // TODO show loading modal ?
+  }
+
+  // Revalidation Effect
+  useEffect(() => {
+    if (
+      fullStructure &&
+      revalidationState.methodToCallOnRevalidation &&
+      (fullStructure?.lastValidationTimestamp || 0) > revalidationState.lastRevalidationRequest
+    ) {
+      const firstInvalidPage = getFirstInvalidPage(fullStructure)
+
+      setRevalidationState({
+        ...revalidationState,
+        methodToCallOnRevalidation: null,
+        shouldProcessValidation: false,
+      })
+      revalidationState.methodToCallOnRevalidation(firstInvalidPage, setStrictSectionPage)
+      // TODO hide loading modal
+    }
+  }, [revalidationState, fullStructure])
+
   console.log('Structure', fullStructure)
```

Comment:
Please remove this one too?

Reply:
goneski: 72d29ae75fe9e477834f98cbdf47aebd9dab83b6

Reply:
😜